tests/stdlib/tjsonutils.nim

```diff
@@ -119,5 +119,95 @@ template fn() =
+    doAssert s == ["one", "two"].toHashSet
+    let jsonNode = toJson(s)
+    doAssert jsonNode.kind == JArray
+    doAssert jsonNode.len == 2
+    let elem0 = jsonNode.elems[0]
+    let elem1 = jsonNode.elems[1]
+    doAssert elem0.kind == JString
+    doAssert elem1.kind == JString
+    doAssert elem0.str == "one" or elem0.str == "two"
+    doAssert elem1.str == "one" or elem1.str == "two"
+    doAssert elem0.str != elem1.str
+
+  block orderedSet:
+    testRoundtrip(["one", "two", "three"].toOrderedSet):
+      """["one","two","three"]"""
+
+  block testJoptions:
+    type
+      AboutLifeUniverseAndEverythingElse = object
+        question: string
+        answer: int
+
+    block testExceptionOnExtraKeys:
+      var guide: AboutLifeUniverseAndEverythingElse
+      let json = parseJson(
+        """{"question":"6*9=?","answer":42,"author":"Douglas Adams"}""")
+      doAssertRaises ValueError, fromJson(guide, json)
+      doAssertRaises ValueError,
+                     fromJson(guide, json, Joptions(allowMissingKeys: true))
+
+      type
+        A = object
+          a1,a2,a3: int
+      var a: A
+      let j = parseJson("""{"a3": 1, "a4": 2}""")
+      doAssertRaises ValueError,
+                     fromJson(a, j, Joptions(allowMissingKeys: true))
+
+    block testExceptionOnMissingKeys:
+      var guide: AboutLifeUniverseAndEverythingElse
+      let json = parseJson("""{"answer":42}""")
+      doAssertRaises ValueError, fromJson(guide, json)
+      doAssertRaises ValueError,
+                     fromJson(guide, json, Joptions(allowExtraKeys: true))
+
+    block testAllowExtraKeys:
+      var guide: AboutLifeUniverseAndEverythingElse
+      let json = parseJson(
+        """{"question":"6*9=?","answer":42,"author":"Douglas Adams"}""")
+      fromJson(guide, json, Joptions(allowExtraKeys: true))
+      doAssert guide.question == "6*9=?"
+      doAssert guide.answer == 42
+
+    block testAllowMissingKeys:
+      var guide: AboutLifeUniverseAndEverythingElse
+      let json = parseJson("""{"answer":42}""")
+      fromJson(guide, json, Joptions(allowMissingKeys: true))
+      doAssert guide.question == ""
+      doAssert guide.answer == 42
+
+    block testAllowExtraAndMissingKeys:
+      var guide: AboutLifeUniverseAndEverythingElse
+      let json = parseJson(
+        """{"answer":42,"author":"Douglas Adams"}""")
+      fromJson(guide, json, Joptions(
+        allowExtraKeys: true, allowMissingKeys: true))
+      doAssert guide.question == ""
+      doAssert guide.answer == 42
+
+    block testExceptionOnMissingDiscriminantKey:
+      type
+        Foo = object
+          a: array[2, string]
+          case b: bool
+          of false: f: float
+          of true: t: tuple[i: int, s: string]
+
+      var foo: Foo
+      let json = parseJson("""{"a":["one","two"]}""")
+      doAssertRaises ValueError, fromJson(foo, json)
+      doAssertRaises ValueError,
+                     fromJson(foo, json, Joptions(allowMissingKeys: true))
+
```

Comment:
* we need 1 more test: when object has an extra field set to a non-default value, it should not be reset to its default value; your tests don't exercise that. eg:
```nim
var guide = AboutLifeUniverseAndEverythingElse(answer = 15)
fromJson(guide, parseJson("""{"question":"foo"}"""), Joptions(allowMissingKeys: true))
doAssert guide == AboutLifeUniverseAndEverythingElse(answer: 15, question: "foo")
```

+ optional nit: consider using single asserts as above instead of 1 assert per field (shorter = better)

Reply:
I fix it and also I added the possibility for optional discriminant keys. You can check the code and the tests for the exact semantic. In short words, the currently set branch is taken when the discriminant key is missing from the JSON.

Reply:
I really want this PR to be merged (it's great work!) and the tests are good now, but something is bothering me with the latest commit: it introduces un-necessary copies with `var newObj: T; fromJsonFields(...); a = newObj` (and possibly extra complexity, but I'm not sure if smthg simpler is possible).

The whole point of `fromJson` was to allow inplace updates whereever possible. But now if even a single field is specified in json with `allowMissingKeys:true`, everything must be copied over still.

Let me think.

Reply:
I introduced the copies intentionally, in order to achieve exception safety. Otherwise if in some moment exception is thrown by the procedure, the object can be left partially filled with old values and partially with the new ones. I many cases this can be completely fine but in some cases, it will be wrong. The choice was between performance and correctness.

But know it came to my mind that this can be better resolved on the caller side and I will follow your advice to avoid the copies when possible.

Reply:
Done.

Reply:
> this can be better resolved on the caller side

indeed; so user can choose between no mutation under exception thrown (edge case) vs performance (common case)